packages/cli/src/server/mod.rs

```diff
@@ -37,95 +50,108 @@ async fn setup_file_watcher<F: Fn() -> Result<BuildResult> + Send + 'static>(
-                    // find changes to the rsx in the file
-                    let mut rsx_file_map = hot_reload.file_map.lock().unwrap();
-                    let mut messages: Vec<Template<'static>> = Vec::new();
-
-                    // In hot reload mode, we only need to rebuild if non-rsx code is changed
-                    needs_full_rebuild = false;
-
-                    for path in &e.paths {
-                        // if this is not a rust file, rebuild the whole project
-                        if path.extension().and_then(|p| p.to_str()) != Some("rs") {
-                            needs_full_rebuild = true;
-                            break;
-                        }
-
-                        // Workaround for notify and vscode-like editor:
-                        // when edit & save a file in vscode, there will be two notifications,
-                        // the first one is a file with empty content.
-                        // filter the empty file notification to avoid false rebuild during hot-reload
-                        if let Ok(metadata) = fs::metadata(path) {
-                            if metadata.len() == 0 {
-                                continue;
-                            }
-                        }
-
-                        match rsx_file_map.update_rsx(path, &config.crate_dir) {
-                            Ok(UpdateResult::UpdatedRsx(msgs)) => {
-                                messages.extend(msgs);
-                                needs_full_rebuild = false;
-                            }
-                            Ok(UpdateResult::NeedsRebuild) => {
-                                needs_full_rebuild = true;
-                            }
-                            Err(err) => {
-                                log::error!("{}", err);
-                            }
-                        }
-                    }
 
-                    if needs_full_rebuild {
-                        // Reset the file map to the new state of the project
-                        let FileMapBuildResult {
-                            map: new_file_map,
-                            errors,
-                        } = FileMap::<HtmlCtx>::create(config.crate_dir.clone()).unwrap();
-
-                        for err in errors {
-                            log::error!("{}", err);
-                        }
-
-                        *rsx_file_map = new_file_map;
-                    } else {
-                        for msg in messages {
-                            let _ = hot_reload.messages.send(msg);
-                        }
-                    }
-                } else {
+        let Ok(e) = info else {
+            return;
+        };
+
+        if chrono::Local::now().timestamp() <= last_update_time {
+            return;
+        }
+
+        let mut needs_full_rebuild = false;
+        let change =
+            futures::executor::block_on(plugins_watched_paths_changed(&e.paths, &config.crate_dir));
```

Comment:
We can use a channel and a seperate task to make this block on shorter: https://github.com/notify-rs/notify/blob/c3929ed114fbb0bc7457a9a498260461596b00ca/examples/async_monitor.rs#L39

Reply:
just importing `block_on` from futures separately cleans it up quite a bit, and it's not as trivial as it should be to just throw the function into a task because lifetimes and futures mix like rocks and glass in a blender